Clean admin

bnmnetp opened the pull request:
* Removing cruft from admin.py 
* using POST with an XMLHttpRequest object causes the server to fail in 3.7 -- convert to jQuery.post
* Write some new style cases for admin !! @bjones1 I actually got a pytest test to work!

bnmnetp commented:
@bjones1 

This test will run and pass when I run it with the complete suite of tests, however when I try to run it standalone with `pytest applications/runestone/tests/test_server.py::test_11` it fails with a validation error.  I put this together looking at other tests and trying to make use of the fixtures and tools in the way I assumed they would work...  Obviously being able to run a single test is critical for getting others to develop tests as waiting for all the tests to run requires more patience than we can reasonably expect.

```
def test_11(test_client, runestone_db_tools, test_user):
    with runestone_db_tools.create_course('test_course_3') as course_3:
        with test_user('test_instructor_1', 'password_1', course_3.course_name) as test_instructor_1, \
            test_instructor_1.make_instructor():

            test_instructor_1.login()
            db = runestone_db_tools.db

            # Create an assignment -- using createAssignment
            test_client.post('admin/createAssignment',
                data=dict(name='test_assignment_1'))

            assign = db((db.assignments.name == 'test_assignment_1') &
                        (db.assignments.course == test_instructor_1.course_id)).select().first()

            assert assign

            # Delete an assignment -- using removeassignment
            test_client.post('admin/removeassign', data=dict(assignid=assign.id))
            assert not db(db.assignments.name == 'test_assignment_1').select().first()

            test_client.post('admin/removeassign', data=dict(assignid=9999999))
            assert "Error" in test_client.text

            test_client.post('admin/removeassign', data=dict(assignid=""))
            assert "Error" in test_client.text
```

I did manage to create a version that will pass when run on its own, but I'd like to understand what I did wrong in the code above, or if there is something in the fixtures that should be "fixed" so that it works the way one might assume.

```
def test_11(test_client, runestone_db_tools, test_user, test_user_1):
    with test_user_1.make_instructor():

        db = runestone_db_tools.db

        # Create an assignment -- using createAssignment
        test_client.post('admin/createAssignment',
            data=dict(name='test_assignment_1'))

        assign = db((db.assignments.name == 'test_assignment_1') &
                    (db.assignments.course == test_user_1.course_id)).select().first()

        assert assign

        # Delete an assignment -- using removeassignment
        test_client.post('admin/removeassign', data=dict(assignid=assign.id))
        assert not db(db.assignments.name == 'test_assignment_1').select().first()

        test_client.post('admin/removeassign', data=dict(assignid=9999999))
        assert "Error" in test_client.text

        test_client.post('admin/removeassign', data=dict(assignid=""))
        assert "Error" in test_client.text
```

bnmnetp commented on controllers/dashboard.py:
```diff
@@ -40,7 +40,7 @@ def index():
     assignments = db(db.assignments.course == course.id).select(db.assignments.ALL, orderby=db.assignments.name)
     logger.debug("getting chapters for {}".format(auth.user.course_name))
-    chapters = db(db.chapters.course_id == auth.user.course_name).select()
+    chapters = db(db.chapters.course_id == course.base_course).select()
```
this is one of those places where the dynamic pages clash with the static pages... To preserve backward compatibility we may want to have a server side config variable

bjones1 replied:
Why run a server with static pages at all? I would contend that fewer options means more testing and more use of the code. Having lots of options / config vars means test case explosion, maintenance burden, and leads to cruft and code bloat. IMHO, supporting too many things means poor support.

So, I'd argue for **only** supporting dynamic routing if the server is in use to keep the code simple and clean. More config vars = more ways to screw up a server.

bnmnetp replied:
I hear the complexity and testability arguments...  

The one big use case I was thinking about was for courses that build for people that are doing highly customized courses using Jenkins.  Whenever they update a Fork of one of the runestone books, Jenkins builds it and deploys it to runestone/static

I suppose that could be done as a dynamic build as well...

bjones1 replied:
(Lots of cheering for fewer options, fewer code paths, better test coverage...)

bjones1 commented:
Your unmodified test code runs for me -- I don't see anything wrong with it, while the new code looks suspicious (no class/instructor created).

bnmnetp commented:
The version I have committed and pushed to the branch the second version in my comment above.  It does create an instructor on the first line of the function. And simply reuses  test_course_1.

bjones1 commented:
Oops, you're correct. I can't reproduce the failure in the first version, and the second version in fact looks OK. I haven't tried this in Docker yet -- does it fail there as well? Perhaps that's a way to reproduce?

bnmnetp commented:
Yes, the first version fails every time in Docker -- unless it is run as part of the entire suite, then it works.

Here is what I get:

```
root@734bce4d35da:/srv/web2py# pytest -v applications/runestone/tests/test_server.py::test_11
======================================= test session starts ========================================
platform linux -- Python 3.7.3, pytest-4.5.0, py-1.8.0, pluggy-0.12.0 -- /usr/local/bin/python
cachedir: .pytest_cache
rootdir: /srv/web2py
collected 1 item

applications/runestone/tests/test_server.py::test_11 FAILED                                  [100%]

============================================= FAILURES =============================================
_____________________________________________ test_11 ______________________________________________

test_client = <applications.runestone.tests.test_server._TestClient object at 0x7f5b042292e8>
runestone_db_tools = <applications.runestone.tests.test_server._RunestoneDbTools object at 0x7f5b04229630>
test_user = <function test_user.<locals>.<lambda> at 0x7f5b04178620>

    def test_11(test_client, runestone_db_tools, test_user):
        with runestone_db_tools.create_course('test_course_3') as course_3:
            with test_user('test_instructor_1', 'password_1', course_3.course_name) as test_instructor_1, \
                test_instructor_1.make_instructor():

>               test_instructor_1.login()

applications/runestone/tests/test_server.py:964:
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _
applications/runestone/tests/test_server.py:416: in login
    _formname='login',
gluon/contrib/webclient.py:75: in post
    self.get(url, cookies=cookies, headers=headers, auth=auth)
gluon/contrib/webclient.py:64: in get
    headers=headers, method='GET')
gluon/contrib/webclient.py:161: in post
    raise error
gluon/contrib/webclient.py:137: in post
    self.response = opener.open(self.url, to_bytes(data))
/usr/local/lib/python3.7/urllib/request.py:531: in open
    response = meth(req, response)
/usr/local/lib/python3.7/urllib/request.py:641: in http_response
    'http', request, response, code, msg, hdrs)
/usr/local/lib/python3.7/urllib/request.py:563: in error
    result = self._call_chain(*args)
/usr/local/lib/python3.7/urllib/request.py:503: in _call_chain
    result = func(*args)
/usr/local/lib/python3.7/urllib/request.py:755: in http_error_302
    return self.parent.open(new, timeout=req.timeout)
/usr/local/lib/python3.7/urllib/request.py:531: in open
    response = meth(req, response)
/usr/local/lib/python3.7/urllib/request.py:641: in http_response
    'http', request, response, code, msg, hdrs)
/usr/local/lib/python3.7/urllib/request.py:563: in error
    result = self._call_chain(*args)
/usr/local/lib/python3.7/urllib/request.py:503: in _call_chain
    result = func(*args)
/usr/local/lib/python3.7/urllib/request.py:755: in http_error_302
    return self.parent.open(new, timeout=req.timeout)
/usr/local/lib/python3.7/urllib/request.py:531: in open
    response = meth(req, response)
/usr/local/lib/python3.7/urllib/request.py:641: in http_response
    'http', request, response, code, msg, hdrs)
/usr/local/lib/python3.7/urllib/request.py:569: in error
    return self._call_chain(*args)
/usr/local/lib/python3.7/urllib/request.py:503: in _call_chain
    result = func(*args)
_ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _ _

self = <urllib.request.HTTPDefaultErrorHandler object at 0x7f5b04194390>
req = <urllib.request.Request object at 0x7f5b04194eb8>
fp = <http.client.HTTPResponse object at 0x7f5b041a7390>, code = 404, msg = 'NOT FOUND'
hdrs = <http.client.HTTPMessage object at 0x7f5b041a76a0>

    def http_error_default(self, req, fp, code, msg, hdrs):
>       raise HTTPError(req.full_url, code, msg, hdrs, fp)
E       urllib.error.HTTPError: HTTP Error 404: NOT FOUND

/usr/local/lib/python3.7/urllib/request.py:649: HTTPError
--------------------------------------- Captured stdout call ---------------------------------------
Changed session ID runestone
------------------------------------- Captured stdout teardown -------------------------------------
Changed session ID runestone

web2py server stdout
--------------------

b'web2py Web Framework\nCreated by Massimo Di Pierro, Copyright 2007-2019\nVersion 2.18.5-stable+timestamp.2019.04.07.21.13.59\nDatabase drivers available: sqlite3, psycopg2, imaplib, pymysql\n\nplease visit:\n\thttp://127.0.0.1:8000/\nuse "kill -SIGTERM 164" to shutdown the web2py server\n\n\n'

web2py server stderr
--------------------

b'web2py.py: warning: --nogui is deprecated, use --no_gui instead\n'
========================================= warnings summary =========================================
gluon/contrib/webclient.py:29
  /srv/web2py/gluon/contrib/webclient.py:29: DeprecationWarning: invalid escape sequence \<
    FORM_REGEX = re.compile('(\<input name\="_formkey" type\="hidden" value\="(?P<formkey>.+?)" \/\>)?\<input name\="_formname" type\="hidden" value\="(?P<formname>.+?)" \/\>')

gluon/contrib/autolinks.py:143
  /srv/web2py/gluon/contrib/autolinks.py:143: DeprecationWarning: VimeoURLOpener style of invoking requests is deprecated. Use newer urlopen functions/methods
    urllib._urlopener = VimeoURLOpener()

-- Docs: https://docs.pytest.org/en/latest/warnings.html
=============================== 1 failed, 2 warnings in 8.04 seconds ===============================
```

bjones1 commented:
Oops again -- I'm not paying enough attention. I don't expect a truly standalone test to work -- the test environment needs a lot of setup per `run_tests.py`. Instead, I use `python run_tests.py -k test_11` to run individual tests.

bnmnetp commented:
Hmm

```
python applications/runestone/tests/test_server.py -k test_11
Traceback (most recent call last):
  File "applications/runestone/tests/test_server.py", line 39, in <module>
    from gluon.contrib.webclient import WebClient
ModuleNotFoundError: No module named 'gluon'
```

bjones1 commented on controllers/admin.py:
```diff
@@ -719,23 +719,31 @@ def deletecourse():
-        session.flash = "Cannot remove assignment with id of {}".format(request.args[0])
-        return
-    db(db.assignments.id == assignment_id).delete()
+        session.flash = "Cannot remove assignment with id of {}".format(request.vars['assignid'])
```
I'm confused. Is this a user-facing endpoint, or something hit by JS? If the first, then return "Success" makes no sense; if the second, this makes no sense.

bnmnetp replied:
It is called from Javascript, but I added the session.flash when I was testing by calling from the browser with a Naked URL.  -- Hey at least it doesn't just crash anymore!!

bjones1 replied:
Progress! IMHO, I'd `return "Cannot remove..."` to be consistent.

bjones1 commented on controllers/default.py:
```diff
@@ -345,6 +348,9 @@ def coursechooser():
 @auth.requires_login()
 def removecourse():
     if settings.academy_mode:
+      if not request.args(0):
```
Shouldn't this check be outside the `if settings.academy_mode:` clause?

bnmnetp replied:
yes

bjones1 commented on static/js/admin.js:
```diff
@@ -1205,18 +1205,20 @@ function remove_assignment() {
-    obj.send(JSON.stringify({ assignid: 'assignmentid' }));
-    obj.onreadystatechange = function () {
-        if (obj.readyState == 4 && obj.status == 200) {
+    var url = '/runestone/admin/removeassign';
```
Doesn't this need a terminating semicolon?

bjones1 replied:
Sorry, this should be for the next line (1209).

bnmnetp replied:
; are not required, and we don't really have great style guidelines or linting of any of our python or javscript.  We can add that to the todo list after we get better test coverage.

bjones1 replied:
Sounds good.

bnmnetp commented:
OK, It would be nice to update run_tests.py so that if you use -k it does not run the older tests.

bjones1 commented on tests/test_server.py:
```diff
@@ -635,9 +636,9 @@ def test_manual(runestone_db_tools, test_user):
-    # FIXME: This produces an exception.
-    #('default/coursechooser', True, 'xxx', 1),
-    #('default/removecourse', True, 'xxx', 1),
+    # TODO: This soesn't really test the body of either of these
```
I agree. This is just a rough test, and should be replaced with detailed tests. A lot of these tests are very surface and need expanding.

bjones1 commented on tests/test_server.py:
```diff
@@ -650,9 +651,10 @@ def test_manual(runestone_db_tools, test_user):
-    # FIXME: This produces an exception.
-    #('dashboard/exercisemetrics', True, 'xxx', 1),
-    #('dashboard/questiongrades', True, 'Gradebook', 1),
+    # TODO: This produces an exception. -- note, this doesn't really test anything about either
```
I would remove "this produces an exception" since you fixed it.

bjones1 commented:
I agree!

bnmnetp commented:
Thanks, I've addressed your comments (or most of them) in the code I just pushed.

Make sure you take a look at the changes to your dynamic pages PR that I accepted as I did add several changes.

Two really important one's to call out:

1.  I made `dynamic_pages` a separate configuration variable so `use_services` does not imply dynamic pages. -- There are good potential use cases for this if someone chooses to write their own backend for books written using Runestone Components.

2. Serving pages dynamically does not imply or require that a user logs in.  I think its important and maybe even required for some schools that have strict privacy rules, to be able to serve the books without requiring any login.

bjones1 commented:
The changes in c5f7a04 look good to me.

1. I don't like this -- again, more code paths, more confusion, more complexity. We need to simplify the config vars and remove this, not add more.
2. I saw that -- it looks fine to me.
